src/mds/MDSDmclockScheduler.cc

```diff
@@ -0,0 +1,891 @@
+  }
+  dout(10) << __func__ << " thread has been joined" << dendl;
+}
+
+void MDSDmclockScheduler::begin_schedule_thread()
+{
+  scheduler_thread = std::thread([this](){process_request();});
+}
+
+void MDSDmclockScheduler::enable_qos_feature()
+{
+  dout(10) << __func__ << dendl;
+
+  default_conf.set_status(true);
+
+  if (!get_mds_is_active()) {
+    dout(0) << " MDS Rank is not active and QoS feature cannot be enabled." << dendl;
+    return;
+  }
+
+  auto sessionmap = get_session_map();
+  if (sessionmap == nullptr) {
+    dout(10) << __func__ << " sessionmap has nullptr" << dendl;
+    return;
+  }
+
+  if (auto it = sessionmap->by_state.find(Session::STATE_OPEN); it != sessionmap->by_state.end()) {
+    for (const auto &session : *(it->second)) {
+      add_session(session);
+    }
+  }
+  if (auto it = sessionmap->by_state.find(Session::STATE_STALE); it != sessionmap->by_state.end()) {
+    for (const auto &session : *(it->second)) {
+      add_session(session);
+    }
+  }
+}
+
+void MDSDmclockScheduler::cancel_inflight_request()
+{
+  dout(10) << __func__ << dendl;
+  std::lock_guard lock(volume_info_lock);
+  std::list<Queue::RequestRef> req_list;
+
+  auto accum_f = [&req_list] (Queue::RequestRef&& r)
+                  {
+                    req_list.push_front(std::move(r));
+                  };
+
+  for (auto it : volume_info_map) {
+    if (it.second.get_inflight_request()) {
+      dmclock_queue->remove_by_client(it.first, true, accum_f);
+    }
+  }
+
+  dout(10) << __func__ << " canceled requests " << req_list.size() << dendl;
+
+  for (auto& it : req_list) {
+    dout(15) << " canceled request volume_id " << it->get_volume_id() << dendl;
+    handle_request_func(it->get_volume_id(), std::move(it), PhaseType::reservation, 1);
+  }
+  ceph_assert(dmclock_queue->empty() == true);
+}
+
+void MDSDmclockScheduler::disable_qos_feature()
+{
+  uint32_t queue_size;
+  bool dmclock_empty;
+
+  dout(10) << __func__ << dendl;
+
+  default_conf.set_status(false);
+
+  do
+  {
+    mds_unlock();
+    queue_cvar.notify_all();
+    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+    queue_size = get_request_queue_size();
+    mds_lock();
+  } while(queue_size);
+
+  do
+  {
+    mds_unlock();
+    cancel_inflight_request();
+    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+    dmclock_empty = dmclock_queue->empty();
+    mds_lock();
+  } while(!dmclock_empty);
+
+  auto sessionmap = get_session_map();
+
+  if (sessionmap == nullptr) {
+    dout(10) << __func__ << " sessionmap has nullptr" <<  dendl;
+    return;
+  }
+
+  if (auto it = sessionmap->by_state.find(Session::STATE_OPEN); it != sessionmap->by_state.end()) {
+    for (const auto &session : *(it->second)) {
+      remove_session(session);
+    }
+  }
+
+  if (auto it = sessionmap->by_state.find(Session::STATE_STALE); it != sessionmap->by_state.end()) {
+    for (const auto &session : *(it->second)) {
+      remove_session(session);
+    }
+  }
```

Comment:
same double-traversal of the sessions, can we traverse it just once?

Reply:
ditto